Vault: PKI External ca docs (#1901)
hashicorp/web-unified-docs: content/vault/v2.x (rc)/content/docs/secrets/pki-external-ca/index.mdx

similar to the [sign](/vault/api-docs/secret/pki#sign-certificate) and
[issue](/vault/api-docs/secret/pki#generate-certificate-and-key) endpoints in the standard PKI
secrets engine. 

You can choose the workflow that best fits your security requirements and operational needs.

### Workflow comparison

| Feature                  | CSR Workflow                                         | Identifier Workflow                           |
|--------------------------|------------------------------------------------------|-----------------------------------------------|
| **Private key location** | Client generates and keeps                           | Vault generates and provides                  |
| **Key transmission**     | Never leaves client                                  | Vault transmits to client after generation    |
| **Complexity**           | Higher - client must generate CSR and associated key | Lower - client only provides identifiers      |
| **Certificate caching**  | Not supported                                        | Vault caches both certificate and private key |

### When to use each workflow

**Use the CSR workflow when:**

> - You have existing key management processes and infrastructure

Reviewer: ```suggestion
- You have existing key management processes and infrastructure.
```

    identifier="www.example.com" \
    challenge_type="http-01"

# Wait for certificate issuance (automatic)
# Check status
$ vault read pki-external-ca/role/web-servers/order/01936d8e-7c3a-7890-b123-456789abcdef

# Retrieve certificate when ready
$ vault read pki-external-ca/role/web-servers/order/01936d8e-7c3a-7890-b123-456789abcdef/fetch-cert
```

### CSR workflow

In the CSR workflow, you generate your own private key and CSR retrieving back only the signed certificate back from Vault:

```shell-session
# Generate private key and CSR locally
$ openssl req -new -newkey rsa:2048 -nodes \
    -keyout server.key \
    -out server.csr \
    -subj "/CN=www.example.com"

# Create order with CSR
$ vault write pki-external-ca/role/web-servers/new-order \
    csr=@server.csr

# Follow same challenge fulfillment process as identifier workflow
# ...

# Retrieve certificate (without private key since you provided it)
$ vault read pki-external-ca/role/web-servers/order/01936d8e-7c3a-7890-b123-456789abcdef/fetch-cert
```

## Role configuration options

Within each role, you can specify which ACME account to use, which domains are allowed,
and how challenges should be fulfilled.

> ### Limiting allowed domains

Reviewer: ```suggestion
### Limit allowed domains
```

influences how Vault matches the allowed_domains against the identifiers in the new-order request.

#### Bare domains

Allow matching domains without any subdomain prefix, such as example.com:

```shell-session
$ vault write pki-external-ca/role/example \
    acme_account_name="letsencrypt-prod" \
    allowed_domains="example.com,www.example.com" \
    allowed_domain_options="bare_domains"

# Allows: example.com, www.example.com
# Denies: api.example.com, sub.example.com
```

#### Subdomains

Allow subdomains of specified domains:

```shell-session
$ vault write pki-external-ca/role/example \
    acme_account_name="letsencrypt-prod" \
    allowed_domains="example.com" \
    allowed_domain_options="subdomains"

# Allows: www.example.com, api.example.com, sub.api.example.com
# Denies: example.com, example.org
```

#### Wildcards

Allow wildcard certificates:

```shell-session
$ vault write pki-external-ca/role/example \
    acme_account_name="letsencrypt-prod" \
    allowed_domains="*.example.com" \
    allowed_domain_options="wildcards" \
    allowed_challenge_types="dns-01"

# Allows: *.example.com
# Denies: example.com, *.*.example.com

> # Note: Wildcard certificates require DNS-01 challenge

Reviewer: ```suggestion
```

```

#### Globs

Allow glob patterns in domain specifications:

```shell-session
$ vault write pki-external-ca/role/example \
    acme_account_name="letsencrypt-prod" \
    allowed_domains="*.prod.example.com,*.staging.example.com" \
    allowed_domain_options="globs"

# Allows: api.prod.example.com, web.staging.example.com
# Denies: example.com, prod.example.com
```

#### Identity templating

> Roles can use ACL path templating to create dynamic domain allowlists based on the authenticated entity:

Reviewer: ```suggestion
Use ACL path templating in the `allowed_domains` field to create dynamic domain
allowlists based on the authenticated entity. For example, the following
configuration allows `<username>.users.example.com` provided the entity name of
the user and the requested subdomain match:
```

```shell-session
$ vault write pki-external-ca/role/user-certs \
    acme_account_name="letsencrypt-prod" \
    allowed_domains="{{identity.entity.name}}.users.example.com" \
    allowed_domain_options="bare_domains"
```

When user "alice" requests a certificate:
- Allowed: `alice.users.example.com`
- Denied: `bob.users.example.com`

Available template variables:
- `{{identity.entity.id}}` - Entity ID
- `{{identity.entity.name}}` - Entity name
- `{{identity.entity.metadata.<key>}}` - Entity metadata
- `{{identity.groups.names.<group>}}` - Group membership

> ### Domain matching behavior

Reviewer: ```suggestion
## Domain matching behavior
```
I suspect folks will reference the provided table frequently so it's probably worth elevating it to a top-level section to make it easier to find (and so it shows up in the TOC)

The following table breaks down how the values within a role's allowed_domains and allowed_domain_options

> interact with each other to forbid or allow identifiers when you call the new-order API.

Reviewer: ```suggestion
The following table breaks down how the `allowed_domains` and
`allowed_domain_options` for a role interact to forbid or allow identifiers when
you call the `new-order` endpooint.
```
Style correction: avoid possessives

- `completed` - Certificate issued successfully
- `expired` - Order expired before completion
- `revoked` - Vault revoked the certificate
- `error` - An error occurred

### List active orders

```shell-session
$ vault list pki-external-ca/role/web-servers/active-orders
```

### Lookup certificate by serial

```shell-session
$ vault read pki-external-ca/lookup/cert/03:e7:1f:a2:3d
```

### View ACME account details

```shell-session
$ vault read pki-external-ca/config/acme-account/letsencrypt-prod
```

## Known limitations

### Load-balanced applications with multiple instances

The current implementation works best when a single orchestrator or instance manages certificate requests.
If you deploy multiple instances of an application behind a load balancer, where each instance independently requests
certificates from Vault, you may encounter challenges:

**The problem:**
- Multiple instances may create separate orders for the same identifiers simultaneously
- Each order generates unique challenge tokens specific to that order
- The ACME CA may validate challenges against a different instance than the one that received the challenge information

> - This can cause validation failures and certificate issuance problems

Reviewer: ```suggestion
- Multiple instances may create separate orders for the same identifiers simultaneously.
- Each order generates unique challenge tokens specific to that order.
- The ACME CA may validate challenges against a different instance than the one
  that received the challenge information and cause validation failures or
  certificate issuance problems.
```
Style correction: avoid "this" as a pronoun

**Recommended workarounds:**

1. **Single orchestrator pattern**: Designate one instance or orchestrator to handle certificate requests. After Vault issues the certificate, distribute it to all instances that need it.

2. **Identifier workflow with caching**: Use the identifier workflow (not CSR workflow) so Vault generates and caches both the certificate and private key. Vault then allows all instances to retrieve the same certificate and key.

### IP address identifiers

The plugin currently does not support requesting certificates with IP address identifiers. 

### ACME Renewal Information (ARI)

The plugin does not currently support the ACME Renewal Information (ARI) extension ([RFC 9773](https://datatracker.ietf.org/doc/rfc9773/)). 

> This extension allows ACME servers to notify clients when they should renew certificates, rather than clients deciding renewal timing independently.

Reviewer: ```suggestion
The plugin does not currently support the ACME Renewal Information (ARI)
extension ([RFC 9773](https://datatracker.ietf.org/doc/rfc9773/)), which allows
ACME servers to notify clients when they should renew certificates. As a result,
your clients need to decide certificate renewal timing independently.
```
Style correction: avoid "this" as a pronoun

secrets engine. 

You can choose the workflow that best fits your security requirements and operational needs.

### Workflow comparison

| Feature | CSR Workflow | Identifier Workflow |
|---------|-------------|---------------------|
| **Private key location** | Client generates and keeps | Vault generates and provides |
| **Key transmission** | Never leaves client | Vault transmits to client after generation |
| **Security model** | Higher - key never transmitted | Standard - Vault transmits key once |
| **Complexity** | Higher - client must generate CSR and associated key | Lower - client only provides identifiers |
| **Certificate caching** | Not supported | Vault caches both certificate and private key |

### When to use each workflow

**Use the CSR workflow when:**
- You require maximum security and want private keys to never leave the client system
- You have existing key management processes and infrastructure
- You need to comply with security policies that prohibit key transmission
- You deploy single-instance applications where key sharing is not needed

**Use the identifier workflow when:**

> - You prioritize operational simplicity over maximum security

Author: Fair I've tweaked and removed the maximum security from both sections.
